readme.md
sec_scan
===============


Ports Scanner for NodeJS. Work in progress.

Status
-------

Currently support :

* individual IP or IP range scan
* individual port, ports list, or ports range
* banner grabbing (not fully implemented, works with verbose ports only)
* IAC negotiation
* reverse dns
* geolocation information
* shell or json output
* optional progress details


Roadmap
--------
* udp scan
* syn scan


Install
-------

>```
>npm install -g sec_scan
>```


Use secscan as a node module
-----------
```
var secscan = require('./');

var options = {
    target:'127.0.0.1',
    port:'21-23',
    status:'TROU', // Timeout, Refused, Open, Unreachable
    banner:true
};

var scanner = new secscan(options);

scanner.on('result',function(data) {
        // fired when item is matching options
            console.log(data);
});

scanner.on('error',function(err) {
        throw new Error(data.toString());
});

scanner.on('done',function() {
        // finished !
});

scanner.run();

```

Use secscan as a command line tool
-----------
Usage: secscan <fqdn|ipv4|cidr> [options]
Example:
```
root@ubuntu:~# secscan 192.168.0.0/24 --port=21-23,80
```


Options
-------
```
  --port          port(s) you want to scan, examples:
                  --port=80
                  --port=21,22
                  --port=21,22,23,5900-5902
                                                                                
  --reverse       display DNS reverse lookup                                    
  
  --reversevalid  only display results having a valid reverse dns, except if
                  ports specified                                               
  
  --geo           display geoip (free maxmind)                                  
  
  --banner        display banner                                                
 
  --bannerlen     set banner length grabing
                  default 512

  --bannerraw     display raw banner (as a JSON Buffer)                         
  
  --progress      display progress indicator each seconds
                                                                                
  --status        ports status wanted in results (example --status=OT)
                  T(timeout)
                  R(refused)
                  O(open, default)
                  U(unreachable)
                                                                                
  --scan          scan method
                  tcpconnect (full connect, default)
                  tcpsyn (half opened, not yet implemented)
                  udp (not yet implemented)
                                                                                
  --concurrency   max number of simultaneous socket opened
                  default 500
                                                                                
  --timeout       maximum number of milliseconds before closing the connection
                  default 2000
                                                                                
  --display       display result format (json,xml,console)
                  default console
                                                                                
  --json          shortcut for --display=json                                   
  
  --xml           shortcut for --display=xml                                    
  
  --console       shortcut for --display=console                                
  
  --help          display help                                                  
  
  --about         display about                                                 
  
  --version       display version number                                        
```

Samples output
----------------

* Every ports on localhost, grab banner, display only opened ports 
```
root@ubuntu:~# secscan 127.0.0.1 --port=0-65535 --banner
127.0.0.1|111||open
127.0.0.1|53||open
127.0.0.1|23|Debian GNU/Linux jessie/sid\r\ndebian login:|open
127.0.0.1|5432||open
127.0.0.1|27017||open
127.0.0.1|28017||open
127.0.0.1|35223||open
127.0.0.1|35491||open
127.0.0.1|39619||open
```

* Every ports on localhost, grab banner, display only opened ports, json output, progress status each seconds
```
root@ubuntu:~# secscan 127.0.0.1 --port=0-65535 --banner --isopen --istimeout --progress --json
{"_timeStart":"N/A","_timeElapsed":"N/A","_jobsTotal":65535,"_jobsRunning":0,"_jobsDone":0,"_progress":0,"_concurrency":500,"_status":"Starting","_message":"Starting"}
{"_timeStart":1371245901876,"_timeElapsed":1031,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":7638,"_progress":11,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:8138"}
{"_timeStart":1371245901876,"_timeElapsed":2085,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":16137,"_progress":24,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:16637"}
{"ip":"127.0.0.1","port":111,"status":"open"}
{"ip":"127.0.0.1","port":53,"status":"open"}
{"ip":"127.0.0.1","port":23,"banner":"Debian GNU/Linux jessie/sid\\r\\ndebian login:","status":"open"}
{"ip":"127.0.0.1","port":5432,"status":"open"}
{"_timeStart":1371245901876,"_timeElapsed":3107,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":24656,"_progress":37,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:25156"}
{"_timeStart":1371245901876,"_timeElapsed":4166,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":33166,"_progress":50,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:33666"}
{"_timeStart":1371245901876,"_timeElapsed":5215,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":41664,"_progress":63,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:42164"}
{"ip":"127.0.0.1","port":27017,"status":"open"}
{"ip":"127.0.0.1","port":28017,"status":"open"}
{"_timeStart":1371245901876,"_timeElapsed":6217,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":49682,"_progress":75,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:50182"}
{"ip":"127.0.0.1","port":35491,"status":"open"}
{"ip":"127.0.0.1","port":35223,"status":"open"}
{"ip":"127.0.0.1","port":39619,"status":"open"}
{"_timeStart":1371245901876,"_timeElapsed":7234,"_jobsTotal":65535,"_jobsRunning":500,"_jobsDone":57732,"_progress":88,"_concurrency":500,"_status":"Running","_message":"Scanning 127.0.0.1:58232"}
{"_timeStart":1371245901876,"_timeElapsed":8182,"_jobsTotal":65535,"_jobsRunning":0,"_jobsDone":65535,"_progress":100,"_concurrency":500,"_status":"Finished","_message":"Scanning 127.0.0.1:65535"}
```

Tips :
--------
**Conncurrency and fast scan**

By default, concurrency is 100. Which is slow when you are scanning large ip range or large port range. You can pass a bigger value using --concurrency option. 1000 is fast by example.
On some linux, only 1024 opened sockets are allowed in the same time. To break this limit, you have to update ulimit parameter of your linux first :

```
ulimit -u unlimited
```

In all cases, you will not be able to scan more than 16580355 ip addr at the moment.


**Pause/unpause**

You can pause/unpause a running scan by sending SIGUSR2 signal. First time it will pause the process, second time it will unpause it.
```
kill -SIGUSR2 19859 # where 19859 is the pid of nodejs process running secscan
```

Help needed
----------
Do not hesitate to open issues and ask for features. Right now i just need "connect" scan. UDP implementation or syn scans should be great. More and more tests.

If you are using secscan, **please star it** !




changelog.md
v1.0.2 - 2017/05/02
===================
* Examples/SaveOutputToFile ==> Save Output in JSON FILE

v1.0.1 - 2017/05/02
===================
* Update Package Name

v1.0.0 - 2017/05/02
===================
* First npm module
* Really bad code

test/A.options.module.js

var expect = require('chai').expect;
var path = require('path');
var secscan = require('../main');

suite(path.basename(__filename), function() {

    var option = require('../libs/options');

    var defaultConcurrency = 500;
    var defaultTimeout = 2000;

    var list = [{
        title:'default args',
        args:{
            target:'127.0.0.1'
        },
        result:{
            "_":[],
            "target":"127.0.0.1",
            "concurrency":500,
            "timeout":2000,
            "status":"O",
            "showOpen":true,
            "scan":"tcpconnect",
            "display":"console",
            "console":true,
            "ips":["127.0.0.1"],
            "ports":[0]
        }
    },{
        title:'all args',
        args:{
            target:'127.0.0.1/29',
            port:'21-23,80',
            status:'TROU',
            concurrency:1,
            timeout:1,
            json:true,
            geo:true,
            reverse:true
        },
        result:{
            "_":[],
            "target":"127.0.0.1/29",
            "port":'21-23,80',
            "status":"TROU",
            "concurrency":1,
            "timeout":1,
            "json":true,
            "geo":true,
            "reverse":true,
            "showTimeout":true,
            "showRefuse":true,
            "showOpen":true,
            "showUnreachable":true,
            "scan":"tcpconnect",
            "display":"json",
            //note that console:true vanish rather than false ..
            "ips":[
                "127.0.0.1",
                "127.0.0.2",
                "127.0.0.3",
                "127.0.0.4",
                "127.0.0.5"
            ],
            "ports":[21,22,23,80]
        }
    }];

    while (list.length) {
        var l = list.shift();
        test(l.title,function(next) {
            var scan = new secscan(l.args,function(s) {
                expect(s.options).to.be.deep.equal(l.result);
                next();
            });
        });
    }

});



test/B.only.dns.single.js
var expect = require('chai').expect;
var net = require('net')
var spawn = require('child_process').spawn;
var rl = require('readline');
var secscan = require('../main.js');
var path = require('path');

var cleanCmdLineArgs = function(str) {
    str = str.replace(/.\/bin\/secscan.js /,' ').trim();
    str = str.replace(/\-\-/g,'');
    return '['+str+']';
}


suite(path.basename(__filename), function() {

    var checkResult = function(data,exp) {
        if (typeof data !='object') {
            data = data.toString();
            expect(data,'response type should be a string').to.be.a('string');
            expect(data.length > 1,'response length should be >1').to.be.ok;
            data = JSON.parse(data);
        }
        expect(data,'JSON.parse should return an object').to.be.a('object');
        expect(data,'object should match').to.be.deep.equal(exp);
        return true;
    }

    var arr = [{
        title:'should return a real result',
        args:'./bin/secscan.js 198.252.206.140 --reverse --json',
        data:{
            ip:'198.252.206.140',
            reverse:"stackoverflow.com"
        }
    }]

    /* simulate command line */

    arr.forEach(function(item) {
        test('Binary: ['+item.args+'] '+item.title,function(next) {

            var checked = false;
            var proc = spawn('node',item.args.split(' '));
            var linereader = rl.createInterface(proc.stdout, proc.stdin);

            linereader.on('line',function(data) {
                if (item.data) return checked = checkResult(data,item.data);
                checked = true;
            });

            proc.on('close',function() {
                expect(checked,'line received before close proc').to.be.ok;
                next();
            });

            proc.stderr.on('data',function(data) {
                throw new Error(data.toString());
            });

        });
    });

    /* simulate module usage */

    arr.forEach(function(item) {
        var o = item.args.split(' ');
        var argv = {};
        argv.target = o[1];
        o.forEach(function(arg) {
            if (arg.match(/\-\-/)) {
                arg = arg.replace(/\-\-/,'');
                if (arg.match(/=/)) {
                    var v = arg.split('=');
                    argv[v[0]]=parseInt(v[1]);
                } else {
                    argv[arg] = true;
                }
            }
        });

        test('Module: '+cleanCmdLineArgs(item.args)+' '+item.title,function(next) {

            this.timeout(5000);
            var checked = false;

            var scan = new secscan(argv,function(s) {

                //console.log('ici',s.options);

                s.on('result',function(data) {
                    expect(data.port).to.be.a('undefined');
                    expect(data.reverse).to.be.a('string');
                    expect(data.reverse.length>0).to.be.ok;
                    if (item.data) return checked = checkResult(data,item.data);
                    checked = true;
                });

                s.on('error',function(err) {
                    throw new Error(data.toString());
                });

                s.on('done',function() {
                    expect(checked,'line received before close proc').to.be.ok;
                    next();
                });

                s.run();
            });
        });
    });
});


test/B.only.geo.single.js

var expect = require('chai').expect;
var net = require('net')
var spawn = require('child_process').spawn;
var rl = require('readline');
var secscan = require('../main.js');
var path = require('path');

var cleanCmdLineArgs = function(str) {
    str = str.replace(/.\/bin\/secscan.js /,' ').trim();
    str = str.replace(/\-\-/g,'');
    return '['+str+']';
}

suite(path.basename(__filename), function() {

    var checkResult = function(data,exp) {
        if (typeof data !='object') {
            data = data.toString();
            expect(data,'response type should be a string').to.be.a('string');
            expect(data.length > 1,'response length should be >1').to.be.ok;
            data = JSON.parse(data);
        }
        expect(data,'JSON.parse should return an object').to.be.a('object');
        expect(data,'object should match').to.be.deep.equal(exp);
        return true;
    }

    var arr = [{
        title:'should return a real result',
        args:'./bin/secscan.js 173.194.45.67 --geo --json',
        data: {
            ip: '173.194.45.67',
            city: 'Mountain View',
            country: 'US',
            region: 'CA',
            latitude: 37.4192,
            longitude: -122.0574
        }
    }]

    /* simulate command line */

    arr.forEach(function(item) {
        test('Binary: ['+item.args+'] '+item.title,function(next) {

            var checked = false;
            var proc = spawn('node',item.args.split(' '));
            var linereader = rl.createInterface(proc.stdout, proc.stdin);

            linereader.on('line',function(data) {
                if (item.data) return checked = checkResult(data,item.data);
                checked = true;
            });

            proc.on('close',function() {
                expect(checked,'line received before close proc').to.be.ok;
                next();
            });

            proc.stderr.on('data',function(data) {
                throw new Error(data.toString());
            });

        });
    });

    /* simulate module usage */

    arr.forEach(function(item) {
        var o = item.args.split(' ');
        var argv = {};
        argv.target = o[1];
        o.forEach(function(arg) {
            if (arg.match(/\-\-/)) {
                arg = arg.replace(/\-\-/,'');
                argv[arg] = true;
            }
        });

        test('Module: '+cleanCmdLineArgs(item.args)+' '+item.title,function(next) {

            var checked = false;

            var scan = new secscan(argv,function(s) {

                s.on('result',function(data) {
                    if (item.data) return checked = checkResult(data,item.data);
                    checked = true;
                });

                s.on('error',function(err) {
                    throw new Error(data.toString());
                });

                s.on('done',function() {
                    expect(checked,'line received before close proc').to.be.ok;
                    next();
                });

                s.run();
            });
        });
    });
});


main.js
var util = require('util');
var events = require('events').EventEmitter;
var tcpconnect = require('./libs/tcpconnect');
var options = require('./libs/options');

var dns = require('dns');
var async = require('async');
var geoip = require('geoip-lite');


var secscan = function(opts,cb) {
    if (false === (this instanceof secscan)) {
        return new secscan(opts,cb);
    }

    var self = this;

    this.lastMessage = 'Starting';
    this.paused = false;
    this.progress = 0;
    this.progressTimer = null;
    this.info = {
        nbIpToScan:0,
        nbPortToScan:0
    };
    this.cacheGeo = {};
    this.cacheDns = {};

    this.options = opts;

    if (!opts.ips) {
        // Called from a js script, reparse options
        options.parse(opts,function(err,opts) {
            self.options = opts;
            self.init();
            events.call(self);
            if (cb) cb(self);
        });
    } else {
        this.init();
        events.call(this);
    }
};

util.inherits(secscan, events);

secscan.prototype.getOptions = function() {
    return this.options;
};

secscan.prototype.initTestLocalhost = function() {
    if (
        this.options.ips.length == 1
        &&
        this.options.ips[0].match(/^127.0.0.1\//)
    ) {
        // For testing purpose, a telnet server can not
        // accept more than 50 simultaneous connection,
        // so let's make a pause between each pools
        this.q.setInterval(1000);
    }
};


secscan.prototype.fireProgress = function() {
    var o = this.q.stats();
    if (!this.paused) {
        o._message = this.lastMessage;
    } else {
        o._message = 'Paused';
        o._status = 'Paused';
    }

    this.progress = o._progress;
    this.emit('progress',o);
};

secscan.prototype.initQueueProgress = function() {
    if (!this.options.progress) return;

    var self = this;

    this.progressTimer = setInterval(self.fireProgress.bind(this),1000);

    this.q.on('start',function() {
        self.emit('start');
        self.fireProgress();
    });

    this.q.on('end',function() {
        clearInterval(self.progressTimer);
        self.fireProgress();
        if (self.cb) self.cb();
    });

};

secscan.prototype.initQueue = function() {

    var self = this;

    // Push scan jobs in the queue
    var maxi = this.options.ips.length;
    var maxj = this.options.ports.length;
    var maxt = maxi*maxj;
    this.info.combinaison = maxt;
    var i = 0;

    var copyOfIps = JSON.parse(JSON.stringify(this.options.ips));

    while (copyOfIps.length) {
        var ip = copyOfIps.shift();
        if (this.options.ports.length) {
            var j = 0;
            var maxj = this.options.ports.length;
            for (var j=0;j<maxj;j++) {
                var args = {
                    ip:ip,
                    port:this.options.ports[j],
                    i:i,
                    max:maxt
                }
                this.info.nbPortToScan++;
                this.q.add(this.scan.bind(this),args);
                i++;
            }
            this.info.nbIpToScan++;
        } else {
            var args = {
                ip:ip,
                i:i,
                max:maxt
            }
            this.q.add(this.scan.bind(this),args);
            this.info.nbPortToScan = 0;
            this.info.nbIpToScan = 1;
            i++;
        }
    }

    this.q.on('end',function() {
        self.cacheGeo = {};
        self.cacheDns = {};
        self.emit('done');
        if (self.cbrun) self.cbrun();
    });

    this.q.on('jobStart',function(data) {
        var str = 'Scanning '+data.ip;
        if (data.port) str+=':'+data.port;
        self.lastMessage = str;
        self.emit('scan',data);
    });

    this.q.on('jobEnd',function(data) {
        var str = 'Scanned '+data.ip;
        if (data.port) str+=':'+data.port;
        self.lastMessage = str;
    });
};

secscan.prototype.pause = function() {
    if (this.paused) return;
    this.paused = true;
    this.q.pause(true);
};

secscan.prototype.unpause = function() {
    if (!this.paused) return;
    this.paused = false;
    this.q.pause(false);
};

secscan.prototype.initQueuePause = function() {
    process.on('SIGUSR2',function() {
        if (!this.paused) {
            this.paused = true;
            this.lastMessage = 'Pause';
        } else {
            this.paused = false
        }
        this.q.pause(this.paused);
    }.bind(this));
};

secscan.prototype.init = function() {

    this.q = require('qjobs')({
        maxConcurrency:this.options.concurrency||500
    });

    this.initTestLocalhost();
    this.initQueue();
    this.initQueueProgress();
    this.initQueuePause();
    //portscan.setSocketTimeout(argv.timeout||1000);
};

secscan.prototype.lookupGeo = function(ip,cb) {

    if (!this.options.geo) {
        return cb();
    }

    if (this.cacheGeo[ip]) {
        return cb(null,this.cacheGeo[ip]);
    }

    var geo = geoip.lookup(ip);
    cb(null, geo);
};

secscan.prototype.lookupDns = function(ip,cb) {
    var self = this;
    if (!this.options.reverse) return cb();

    if (this.cacheDns[ip]) {
        return cb(null,this.cacheDns[ip]);
    }

    dns.reverse(ip,function(err,domains) {
        if (err) {
            if (err.code == 'ENOTFOUND') {
                return cb(null);
            } else {
                // unknow error
                var e = {
                    fnc:'lookupDns',
                    err:err
                }
                self.emit('error',e);
                return cb(null);
            }
        }
        if (!domains.length) return cb(null);
        cb(null,domains[0]);
    });
};

secscan.prototype.portScan = function(ip,port,cb) {
    if (!port) return cb();
    if (port == 0) return cb();

    var args = {
        ip : ip,
        port : port,
        banner : this.options.banner,
        bannerlen : this.options.bannerlen,
        timeout : this.options.timeout
    };

    var t = new tcpconnect(args);
    t.analyzePort(cb);
};

secscan.prototype.resultAddGeo = function(result,r) {
    if (!this.options.geo) return r;

    r.city = '';
    r.country = '';
    r.region = '';
    r.latitude = '';
    r.longitude = '';

    if (!result) return r;

    this.cacheGeo[r.ip] = result;

    r.city = result.city || '';
    r.country = result.country || '';
    r.region = result.region || '';
    r.latitude = result.ll[0] || '';
    r.longitude = result.ll[1] || '';

    return r;
};

secscan.prototype.resultAddDns = function(result,r) {
    if (!this.options.reverse || !r) return r;
    r.reverse = '';
    if (!result) return r;
    r.reverse = result;

    this.cacheDns[r.ip] = result;

    return r;
};

secscan.prototype.resultAddPort = function(result,r) {

    if (!r || !this.options.port || !result) {
        return r;
    }

    if (!this.options.showTimeout && result.status.match(/timeout/i)) {
        return r = null;
    }

    if (!this.options.showRefuse && result.status.match(/refuse/i)) {
        return r = null;
    }

    if (!this.options.showOpen && result.status.match(/open/i)) {
        return r = null;
    }

    if (!this.options.showUnreachable && result.status.match(/unreachable/i)) {
        return r = null;
    }

    if (this.options.banner) {
        r.banner = result.banner || '';
    }

    if (this.options.bannerraw) {
        r.bannerraw = result.raw || '';
    }

    r.status = result.status;

    return r;
};

secscan.prototype.resultClean = function(r) {
    if (!r) {
        return r;
    }

    if (this.options.reverse && this.options.reversevalid && r.reverse == '') {
        r = null;
    }

    if (!r.port) delete r.port;

    if (this.options.json) {

        if (r.status && r.status.match(/close/i)) {
            delete r.banner;
            delete r.bannerraw;
        }

        if (!r.reverse) {
            delete r.reverse;
        }

        if (!r.banner) {
            delete r.banner;
        }

        if (!r.bannerraw) {
            delete r.bannerraw;
        }
    }
    return r;
};

secscan.prototype.scan = function(args,nextIteration) {

    var self = this;

    var result = {
        ip:args.ip,
        port:args.port
    };

    async.series([
        function(next) {
            self.lookupGeo(args.ip,next);
        },
        function(next) {
            self.lookupDns(args.ip,next);
        },
        function(next) {
            self.portScan(args.ip,args.port,next);
        }
    ],function(err,arr) {
        result = self.resultAddGeo(arr[0],result);
        result = self.resultAddDns(arr[1],result);
        result = self.resultAddPort(arr[2],result);
        result = self.resultClean(result);
        if (result) {
            self.emit('result',result);
        }
        nextIteration();
    });
};

secscan.prototype.run = function(cb) {
    this.cbrun = cb;
    this.q.run();
    return;
};

secscan.prototype.abort = function() {
    this.q.abort();
    return;
};

secscan.prototype.getInfo = function() {
    return this.info;
};

module.exports = secscan;


libs/options.js
var net = require('net');
var cidr = require('./cidr');
var dns = require('dns');
var async = require('async');
var findup = require('findup-sync');

var getTargets = function(target,cb) {

    if (!target) {
        return cb("Please specify at least a target [cidr|ipv4|host], example:\nsecscan 192.168.0.0/24 --port=21,22,23,80,5900-5910");
    }

    var ips = [];

    async.series([
        function(next) {
            if (target.match(/[a-z]/i) && !target.match(/\//) && !net.isIPv6(target)) {
                dns.resolve4(target,next);
            } else {
                next(null,[[target]]);
            }
        }
    ],function(err,result) {
        if (err) {
            if (err.code=='ENOTFOUND') {
                return cb('Could not resolve '+target);
            }
            return cb(err);
        }

        target = result[0][0]+'';

        if (target.match(/\-/)) {
            var splitTarget = target.split('-'),
                minHost     = splitTarget[0],
                ips         = [],
                splitMinHost, maxHost;

            if (net.isIPv4(minHost)) {
                splitMinHost = minHost.split('.');
                if (net.isIPv4(splitTarget[1])) {
                    maxHost = splitTarget[1].split('.')[3];
                } else {
                    // Check if the string is a positive integer
                    if (splitTarget[1] >>> 0 === parseFloat(splitTarget[1])) {
                        maxHost = splitTarget[1];
                    } else {
                        return cb("Invalid IPv4 target range, ie: 192.168.0.1-5, 192.168.0.1-192.168.0.5");
                    }
                }
            } else {
                return cb("Invalid IPv4 target. ie: 192.168.0.1-5, 192.168.0.1-192.168.0.5");
            }

            for (i = parseInt(splitMinHost[3]); i <= parseInt(maxHost); i++) {
                ips.push(splitMinHost[0] + '.' + splitMinHost[1] + '.' +
                         splitMinHost[2] + '.' + i);
            }

            if (!ips) {
                return cb("Invalid IPv4 target. Please specify a target using --target [cidr|ip|range]");
            }
            return cb(null,ips);
        }

        if (target.match(/\//)) {
            var ips = cidr.get(target);
            if (!ips) {
                return cb("Invalid IPv4 CIDR target. Please specify a target using --target [cidr|ip|range]");
            }
            return cb(null,ips);
        }

        if (net.isIPv6(target)) {
            return cb("IPv6 not supported");
        }

        if (target == '127.0.0.1') {
            return cb(null,[target]);
        }

        if (!net.isIPv4(target)) {
            return cb("Target "+target+" is not a valid IPv4");
        } else {
            return cb(null,[target]);
        }

        return cb("Target: unknow error");
    });
};


var addPortRange = function(range,ports) {
    if (!range.match(/[0-9]+\-[0-9]+/)) return;
    var sp = range.split('-');
    var start = parseInt(sp[0]);
    var end = parseInt(sp[1]);
    if (start+1 && end+1) {
        if (start == 0) start++;
        for (var i = start;i<=end;i++) {
            ports.push(i);
        }
    }
    return true;
};

var getPorts = function(port,cb) {
    var ports = [];

    if (!port) {
        return cb(null,[0]);
    }

    port+='';
    if (port.match(/^[0-9]+$/)) {
        return cb(null,[parseInt(port)]);
    }

    if (!port.match(/[0-9,\-]+/)) {
        return cb("Invalid port "+port);
    }

    port+=',';

    var p = port.split(',');
    p.forEach(function(port) {
        if (!port) return;
        if (!addPortRange(port,ports)) {
            ports.push(parseInt(port));
        }
    });

    if (ports.length) return cb(null,ports);

    return cb('Port: unknow error');
};

var defaultValues = function(argv) {
    if (!argv.concurrency) {
        argv.concurrency = 500;
    }

    if (!argv.timeout) {
        argv.timeout = 2000;
    }

    if (!argv.status) {
        argv.status = 'O';
    }

    if (argv.status.match(/T/)) {
        argv.showTimeout = true;
    }

    if (argv.status.match(/R/)) {
        argv.showRefuse = true;
    }

    if (argv.status.match(/O/)) {
        argv.showOpen = true;
    }

    if (argv.status.match(/U/)) {
        argv.showUnreachable = true;
    }

    if (!argv.scan) {
        argv.scan = 'tcpconnect';
    }

    if (argv.json) {
        argv.display = 'json';
    }

    if (argv.xml) {
        argv.display = 'xml';
    }

    if (argv.console) {
        argv.display = 'console';
    }

    if (!argv.display) {
        argv.display = 'console';
    }

    if (argv.display == 'console') {
        argv.console = true;
    }

    if (argv.display == 'json') {
        argv.json = true;
    }

    if (argv.display == 'xml') {
        argv.xml = true;
    }

    if (!argv.timeout) {
        argv.timeout = 2000;
    }
    return argv;
};

var help = function(optimist,argv) {
    if (argv.help) {
        optimist.showHelp();
        process.exit(0);
    }

    if (argv.version||argv.about) {
        var fs = require('fs');
        var package = JSON.parse(fs.readFileSync(findup('package.json')));
    }

    if (argv.version) {
        console.log(package.version);
        process.exit(0);
    }

    if (argv.about) {
        console.log(
            package.name,
            package.version,'\n',
            'Resume: '+package.description,'\n',
            'License: '+package.license,'\n',
            'Author: '+package.author,'\n',
            'Repository: '+package.repository.url.replace(/git/,'http')
        );
        process.exit(0);
    }
    return argv;
};

var takeCareOfCrazyPeople = function(argv,cb) {
    var count = argv.ips.length * argv.ports.length;
    if (count>16580355) {
        var msg = 'limit of 16580355 numbers of ip/port combinaison reached ('+count+')';
        cb(msg);
        return false;
    }
    return true;
};

var parse = function(args,cb) {

    var optimist = require('optimist')
        .usage('Usage: secscan <fqdn|ipv4|cidr> [options]\n\nExample: secscan 192.168.0.0/24 --port=21-23,80')
        .demand('_')

        .describe(
            'port',
            'port(s) you want to scan, examples:\n'+
            '--port=80\n'+
            '--port=21,22\n'+
            '--port=21,22,23,5900-5900\n'
        )
        .describe(
            'reverse',
            'display DNS reverse lookup'
        )
        .describe(
            'reversevalid',
            'only display results having a valid reverse dns, except if ports specified'
        )
        .describe(
            'geo',
            'display geoip (free maxmind)'
        )
        .describe(
            'banner',
            'display grabbed banner when available'
        )
        .describe(
            'bannerraw',
            'display raw banner (as a JSON Buffer)'
        )
        .describe(
            'bannerlen',
            'grabbed banner length in bytes\n'+
            'default 512'
        )
        .describe(
            'progress',
            'display progress indicator each seconds\n'
        )
        .describe(
            'status',
            'ports status wanted in results (example --status=OT)\n'+
            'T(timeout)\n'+
            'R(refused)\n'+
            'O(open, default)\n'+
            'U(unreachable)\n'
        )
        .describe(
            'scan',
            'scan method\n'+
            'tcpconnect (full connect, default)\n'+
            'tcpsyn (half opened, not yet implemented)\n'+
            'udp (not yet implemented)\n'
        )
        .describe(
            'concurrency',
            'max number of simultaneous socket opened\n'+
            'default 500\n'
        )
        .describe(
            'timeout',
            'maximum number of milliseconds before closing the connection\n'+
            'default 2000\n'
        )
        .describe(
            'hugescan',
            'allow number of ip/port combinaison greater than 16580355\n'+
            '(i.e a /24 network with port range 0-65535)'
        )
        .describe(
            'display',
            'display result format (json,xml,console)\n'+
            'default console\n'
        )
        .describe(
            'json',
            'shortcut for --display=json'
        )
        .describe(
            'xml',
            'shortcut for --display=xml'
        )
        .describe(
            'console',
            'shortcut for --display=console'
        )
        .describe(
            'help',
            'display help'
        )
        .describe(
            'about',
            'display about'
        )
        .describe(
            'version',
            'display version number'
        )
        .wrap(80);

    var argv = optimist.parse(args);

    // merge options when used in a node module
    // because we are passing options without "--"
    // like when using secscan with the command line
    for (var attr in args) {
        argv[attr] = args[attr];
    }

    // we don't care about that
    delete argv['$0'];

    argv = help(optimist,argv);
    argv = defaultValues(argv);

    async.series([
        function(next) {
            getTargets(argv._[2]||args.target,next);
        },
        function(next) {
            getPorts(argv.port||args.port,next);
        }
    ],function(err,result) {

        if (err) return cb(err);

        argv.ips = result[0];
        argv.ports = result[1];

        if (!argv.port && !argv.reverse && !argv.geo) {
            var msg = 'Please specify at least one port, --port=80';
            return cb(msg);
        }

        takeCareOfCrazyPeople(argv,cb);

        cb(null,argv);
    });
};


module.exports = {
    getTargets:getTargets,
    getPorts:getPorts,
    parse:parse
};
